test/libyul/EVMDialectCompatibility.cpp

```diff
@@ -0,0 +1,82 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#include <test/libyul/EVMDialectCompatibility.h>
+
+#include <boost/test/data/test_case.hpp>
+#include <boost/test/data/monomorphic.hpp>
+
+namespace bdata = boost::unit_test::data;
+
+using namespace solidity;
+using namespace solidity::yul;
+using namespace solidity::yul::test;
+
+BOOST_AUTO_TEST_SUITE(EVMDialectCompatibility)
+
+BOOST_DATA_TEST_CASE(
+	builtin_function_handle_compatibility_non_eof,
+	bdata::make(generateEVMDialectConfigurationsToTest(std::nullopt)),
+	evmDialectConfigurationToTest
+)
+{
+	auto const& dialectToTestAgainst = evmDialectConfigurationToTest.dialect();
+	// no object access for current
+	{
+		auto const& currentDialect = EVMDialect::strictAssemblyForEVM({}, std::nullopt);
+		for (auto const& builtinFunctionName: currentDialect.builtinFunctionNames())
+			requireBuiltinCompatibility(currentDialect, dialectToTestAgainst, builtinFunctionName);
+	}
+	// object access for current
+	{
+		auto const& currentDialect = EVMDialect::strictAssemblyForEVMObjects({}, std::nullopt);
+		for (auto const& builtinFunctionName: currentDialect.builtinFunctionNames())
+			requireBuiltinCompatibility(currentDialect, dialectToTestAgainst, builtinFunctionName);
+	}
+}
+
+BOOST_DATA_TEST_CASE(
+	builtin_function_handle_compatibility_eof,
+	bdata::monomorphic::grid(
+		bdata::make(generateEVMDialectConfigurationsToTest(std::nullopt)) + bdata::make(generateEVMDialectConfigurationsToTest(1)),
```

Comment:
This seems to actually be testing both EOF and non-EOF despite the name.

Reply:
And a much more relevant difference is that it's using the latest rather than the default EVM version. In not so distant future both current and latest will have EOF, but the fact that current is not always the latest will remain. If you replace `withEOF` check with `withEOF && evmVersion.supportsEOF()` it will be more future-proof. You can then unify the tests and make the source EVM version just another dimension.

Reply:
Also, I think we should have test case where we cover the "inline" -> "with objects" conversion for all dialects. Converting between distinct dialects is nice to have but these pairs are the most relevant, because for them we actually do perform such a conversion in practice.